src/core/src/atomic_guard.hpp

```diff
@@ -12,7 +12,7 @@ class AtomicGuard {
 public:
     AtomicGuard(std::atomic_bool& b) : m_atomic(b) {
         bool exp = false;
-        while (!m_atomic.compare_exchange_strong(exp, true)) {
+        while (m_atomic.load(std::memory_order_relaxed) || !m_atomic.compare_exchange_strong(exp, true)) {
```

Comment:
Per my understanding, 'memory_order_relaxed' doesn't guarantee to get real value of atomic variable - it can be obtained from current processor's cache. There is no guarantee that this cache will be ever refreshed with real value from RAM, so in theory it can cause the infinite loop.
Thread 1: sets `b=true`
Thread 2: reads 'b' from RAM, puts it's value to cache, and starts waiting loop
Thread 1: sets 'b=false' in destructor, updates value in memory, but not in cache of Thread 2
Thread 2: doesn't take atomic value from RAM (there is no need for this), uses it from cache - infinite loop

If we get into infinite loop or not depends on processor and compiler, which is not safe

A long time ago, before C++11 'volatile' keyword has been used to avoid such 'caching' of variables, now it is not a recommended way.

My suggestion is to change 'memory_order_relaxed' to 'memory_order_seq_cst' or to 'memory_order_acquire'